src/util/uploadExternalFormatFile.tsx

```diff
@@ -0,0 +1,156 @@
+export interface UploadExternalFormatFileFormValues {
+  imageFile: File | null;
+  name: string;
+  pool: string;
+  formatConversion: boolean;
+  virtioConversion: boolean;
+  architecture: string;
+}
+
+export const supportedVMArchOptions = (hostArchitectures: string[]) => {
+  const vmArchitectureOptions = [
+    { value: "x86_64", label: "x86_64" },
+    { value: "aarch64", label: "aarch64" },
+    { value: "ppc64le", label: "ppc64le" },
+    { value: "s390x", label: "s390x" },
+  ];
+
+  return vmArchitectureOptions.filter((arch) =>
+    hostArchitectures.includes(arch.value),
+  );
+};
+
+export const uploadExternalFormatFilePayload = (
+  values: UploadExternalFormatFileFormValues,
+) => {
+  const fileSize = values.imageFile?.size || 0;
+  const conversionOptions = [];
+
+  if (values.formatConversion) {
+    conversionOptions.push("format");
+  }
+
+  if (values.virtioConversion) {
+    conversionOptions.push("virtio");
+  }
+
+  const instanceArgs = {
+    source: {
+      type: "conversion",
+      mode: "push",
+      conversion_options: conversionOptions,
+    },
+    devices: {
+      root: {
+        path: "/",
+        type: "disk",
+        pool: values.pool,
+      },
+    },
+    type: "virtual-machine",
+    name: values.name,
+    sourceDiskSize: fileSize,
+    architecture: values.architecture,
+  };
+
+  return JSON.stringify(instanceArgs);
+};
+
+// This method facilitates the uploading of a file to a server via a websocket connection.
+// 1. When the websocket connection opens, a FileReader instance is created to read the file in chunks.
+// 2. The readNextFileChunk function slices the file into chunks of the specified chunk size and converts them into binary buffers.
+// 3. If the chunk size is zero, indicating the end of the file, the websocket connection is closed.
+// 4. The FileReader.onload event handler processes indiviual file chunks as they get read, which sends them
+// through the websocket connection.
+export const sendFileByWebSocket = (
+  socketURL: string,
+  file: File | null,
+  chunkSize: number,
+  onProgress?: (current: number, total: number) => void,
+  onError?: (error: Error) => void,
+) => {
+  const ws = new WebSocket(socketURL);
+
+  ws.onopen = () => {
+    const reader = new FileReader();
+    let sentFileSize = 0;
+
+    const readNextFileChunk = () => {
+      if (!file) {
+        return;
+      }
+      const chunk = file.slice(sentFileSize, sentFileSize + chunkSize);
+      if (!chunk.size) {
+        // NOTE: if we close the connection without a code, the default 1005 is used which will cause backend to throw an error
+        ws.close(1000, "File upload complete");
+        return;
+      }
+
+      reader.readAsArrayBuffer(chunk);
+    };
+
+    reader.onload = (event) => {
+      if (!file) {
+        return;
+      }
```

Comment:
Can this guard be at the top of the function, so we have it only once?

Reply:
unfortunately not, the external guard does not apply inside the context of the callbacks